File: app/src/main/java/ru/nikitazhelonkin/coinbalance/data/api/exchange/BinanceApiClient.java
package ru.nikitazhelonkin.coinbalance.data.api.exchange;


import java.util.HashMap;

import io.reactivex.Single;
import ru.nikitazhelonkin.coinbalance.data.api.HttpErrorTransformer;
import ru.nikitazhelonkin.coinbalance.data.api.response.BinanceBalancesResponse;
import ru.nikitazhelonkin.coinbalance.data.api.exchange.BinanceApiService;
import ru.nikitazhelonkin.coinbalance.utils.DigestUtil;

public class BinanceApiClient implements ExchangeApiClient {

    private BinanceApiService mApiService;

    public BinanceApiClient(BinanceApiService apiService) {
        mApiService = apiService;
    }

    @Override
    public Single<HashMap<String, Float>> getBalances(String apiKey, String apiSecret) {
        long time = mApiService.time().blockingGet().getServerTime();
        String timeStamp = String.valueOf(time);
        String queryString = String.format("timestamp=%s",  timeStamp);
        String signature = DigestUtil.hmacString(queryString, apiSecret, "HmacSHA256");
        return mApiService.balances(timeStamp, signature, apiKey).map(BinanceBalancesResponse::getNonZeroBalances)
                .compose(new HttpErrorTransformer<>());
    }
}

File: app/src/main/java/ru/nikitazhelonkin/coinbalance/data/api/exchange/BinanceApiClient.kt
package ru.nikitazhelonkin.coinbalance.data.api.exchange


import io.reactivex.Single
import ru.nikitazhelonkin.coinbalance.data.api.HttpErrorTransformer
import ru.nikitazhelonkin.coinbalance.utils.DigestUtil
import java.util.*

class BinanceApiClient(private val apiService: BinanceApiService) : ExchangeApiClient {

    override fun getBalances(apiKey: String, apiSecret: String): Single<HashMap<String, Float>> {
        val time = apiService.time().blockingGet().serverTime
        val timeStamp = time.toString()
        val queryString = String.format("timestamp=%s", timeStamp)
        val signature = DigestUtil.hmacString(queryString, apiSecret, "HmacSHA256")
        return apiService.balances(timeStamp, signature, apiKey)
                .map { it.nonZeroBalances }
                .compose(HttpErrorTransformer())
    }
}
